src/coordinator/api/v1/handler/prometheus/native/read.go

```diff
@@ -114,89 +104,94 @@ func (h *PromReadHandler) read(reqCtx context.Context, w http.ResponseWriter, pa
-		// TODO(nikunj): Use binary search once blocklist becomes big
-		if blockSeries.Meta.Bounds.Start.Before(s.Meta.Bounds.Start) {
-			blockList = append(blockList, block.Series{})
-			copy(blockList[idx+1:], blockList[idx:])
-			blockList[idx] = blockSeries
-			seriesMap[seriesID] = blockList
-			return
-		}
+	firstBlock := blockList[0]
+	numSeries := firstBlock.Series()
+	seriesList := make([]*ts.Series, numSeries)
+	seriesIters := make([]block.SeriesIter, len(blockList))
+	for i, b := range blockList {
+		seriesIters[i] = b.SeriesIter()
 	}
 
-	// If all start times lesser, then append to the end
-	seriesMap[seriesID] = append(blockList, blockSeries)
-}
+	seriesMeta := firstBlock.SeriesMeta()
+	bounds := firstBlock.Meta().Bounds
+	for i := 0; i < numSeries; i++ {
+		values := ts.NewFixedStepValues(bounds.StepSize, firstBlock.Steps()*len(blockList), math.NaN(), bounds.Start)
+		valIdx := 0
+		for idx, iter := range seriesIters {
+			if !iter.Next() {
+				return nil, fmt.Errorf("invalid number of datapoints for series: %d, block: %d", i, idx)
+			}
 
-func seriesMapToSeriesList(seriesMap map[string][]block.Series) ([]ts.Series, error) {
-	seriesList := make([]ts.Series, 0, len(seriesMap))
-	for _, blockSeriesList := range seriesMap {
-		s, err := blockSeriesListToSeries(blockSeriesList)
-		if err != nil {
-			return nil, err
+			blockSeries := iter.Current()
+			for i := 0; i < blockSeries.Len(); i++ {
+				values.SetValueAt(valIdx, blockSeries.ValueAtStep(i))
+				valIdx++
+			}
 		}
 
-		seriesList = append(seriesList, *s)
+		seriesList[i] = ts.NewSeries(seriesMeta[i].Name, values, seriesMeta[i].Tags)
+
 	}
 
 	return seriesList, nil
 }
 
-func blockSeriesListToSeries(series []block.Series) (*ts.Series, error) {
-	if len(series) == 0 {
-		return &ts.Series{}, nil
+func insertSortedBlock(b block.Block, blockList []block.Block) ([]block.Block, error) {
+	if len(blockList) == 0 {
+		blockList = append(blockList, b)
+		return blockList, nil
 	}
 
-	firstSeries := series[0]
-	var totalDatapoints int
-	for _, s := range series {
-		totalDatapoints += s.Len()
+	firstBlock := blockList[0]
+	if firstBlock.Series() != b.Series() {
+		return nil, fmt.Errorf("mismatch in number of series for the block, wanted: %d, found: %d", firstBlock.Series(), b.Series())
 	}
 
-	values := ts.NewFixedStepValues(firstSeries.Meta.Bounds.StepSize, totalDatapoints, math.NaN(), firstSeries.Meta.Bounds.Start)
-	valIdx := 0
-	for _, s := range series {
-		for idx := 0 ; idx < s.Len(); idx++ {
-			values.SetValueAt(valIdx, s.ValueAtStep(idx))
-			valIdx++
-		}
+	if firstBlock.Steps() != b.Steps() {
+		return nil, fmt.Errorf("mismatch in number of steps for the block, wanted: %d, found: %d", firstBlock.Steps(), b.Steps())
 	}
 
-	return ts.NewSeries(firstSeries.Meta.Name, values, nil), nil
+	index := sort.Search(len(blockList), func(i int) bool { return blockList[i].Meta().Bounds.Start.Before(b.Meta().Bounds.Start) })
+	blockList = append(blockList, b)
```

Comment:
I think we can just do `blockList = blockList[0:len(blockList)+1]` instead of appending `b` here

Reply:
you need a way to increase the size of the slice